### PR TITLE
Improve error messages for `Accuracy` interfaces

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

- adds a MisConfigurationError to utilities
- uses MisConfigurationError to provide users a configuration error message in the functional and module interface of `accuracy` if the user passes in an incorrect `task`.
- removed asserts in each interface and replaced with MisConfigurationError

The PR concept was discussed in Slack/new-contributors and acknowledged by Jirka.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by Borda** [comment]
fyi, pre-commit will be fixed in #1464

File: src/torchmetrics/classification/accuracy.py [hunk 1]
```diff
@@ -450,18 +451,31 @@ def __new__(
         validate_args: bool = True,
         **kwargs: Any,
     ) -> Metric:
+
         kwargs.update(
             {"multidim_average": multidim_average, "ignore_index": ignore_index, "validate_args": validate_args}
         )
+
+        if task not in ["binary", "multiclass", "multilabel"]:
```
Comment (Borda) [inline src/torchmetrics/classification/accuracy.py:459]: let's wait for #1479 is merged and use this Enum

Comment (JustinGoheen) [reply src/torchmetrics/classification/accuracy.py:501]: I see that 1479 is merged. Is there anything else I need to do?

Comment (Borda) [reply src/torchmetrics/classification/accuracy.py:501]: Can you please use these enums?

Comment (JustinGoheen) [reply src/torchmetrics/classification/accuracy.py]: implemented in https://github.com/Lightning-AI/metrics/pull/1448/commits/a1e118fd894a928d0ee60d4aca38db8a8477c9d1

File: src/torchmetrics/classification/accuracy.py [hunk 2]
```diff
@@ -493,15 +494,26 @@ def __new__(
...
             {"multidim_average": multidim_average, "ignore_index": ignore_index, "validate_args": validate_args}
         )
+
         if task == ClassificationTask.BINARY:
             return BinaryAccuracy(threshold, **kwargs)
         if task == ClassificationTask.MULTICLASS:
-            assert isinstance(num_classes, int)
-            assert isinstance(top_k, int)
+            if not isinstance(num_classes, int):
+                raise MisConfigurationError(
```
Comment (Borda) [inline src/torchmetrics/classification/accuracy.py:506]:
```suggestion
                raise ValueError(
```

File: src/torchmetrics/classification/accuracy.py [hunk 3]
```diff
@@ -493,15 +494,26 @@ def __new__(
...
+                    f"Optional arg `num_classes` must be type `int` when task is {task}. Got {type(num_classes)}"
+                )
+            if not isinstance(top_k, int):
+                raise MisConfigurationError(
```
Comment (Borda) [inline src/torchmetrics/classification/accuracy.py:510]:
```suggestion
                raise ValueError(
```

File: src/torchmetrics/classification/accuracy.py [hunk 4]
```diff
@@ -493,15 +494,26 @@ def __new__(
...
+                    f"Optional arg `top_k` must be type `int` when task is {task}. Got {type(top_k)}"
+                )
             return MulticlassAccuracy(num_classes, top_k, average, **kwargs)
         if task == ClassificationTask.MULTILABEL:
-            assert isinstance(num_labels, int)
+            if not isinstance(num_labels, int):
+                raise MisConfigurationError(
```
Comment (Borda) [inline src/torchmetrics/classification/accuracy.py:516]:
```suggestion
                raise ValueError(
```

File: src/torchmetrics/functional/classification/accuracy.py [hunk 5]
```diff
@@ -398,18 +399,26 @@ def accuracy(
...
     task = ClassificationTask.from_str(task)
-    assert multidim_average is not None
+
     if task == ClassificationTask.BINARY:
         return binary_accuracy(preds, target, threshold, multidim_average, ignore_index, validate_args)
     if task == ClassificationTask.MULTICLASS:
-        assert isinstance(num_classes, int)
-        assert isinstance(top_k, int)
+        if not isinstance(num_classes, int):
+            raise MisConfigurationError(
```
Comment (Borda) [inline src/torchmetrics/functional/classification/accuracy.py:407]:
```suggestion
            raise ValueError(
```

File: src/torchmetrics/functional/classification/accuracy.py [hunk 6]
```diff
@@ -398,18 +399,26 @@ def accuracy(
...
+                f"Optional arg `num_classes` must be type `int` when task is {task}. Got {type(num_classes)}"
+            )
+        if not isinstance(top_k, int):
+            raise MisConfigurationError(
```
Comment (Borda) [inline src/torchmetrics/functional/classification/accuracy.py:411]:
```suggestion
            raise ValueError(
```

File: src/torchmetrics/functional/classification/accuracy.py [hunk 7]
```diff
@@ -398,18 +399,26 @@ def accuracy(
...
+                f"Optional arg `top_k` must be type `int` when task is {task}. Got {type(top_k)}"
+            )
         return multiclass_accuracy(
             preds, target, num_classes, average, top_k, multidim_average, ignore_index, validate_args
         )
     if task == ClassificationTask.MULTILABEL:
-        assert isinstance(num_labels, int)
+        if not isinstance(num_labels, int):
+            raise MisConfigurationError(
```
Comment (Borda) [inline src/torchmetrics/functional/classification/accuracy.py:419]:
```suggestion
            raise ValueError(
```

File: tests/unittests/classification/test_accuracy.py [hunk 8]
```diff
@@ -60,6 +66,53 @@ def _sklearn_accuracy_binary(preds, target, ignore_index, multidim_average):
...
 
 
+class TestAccuracy(MetricTester):
+    def test_accuracy(self):
+        preds, target = _input_binary
+        task = "NotValidTask"
+        ignore_index = None
+        multidim_average = "global"
+
+        with pytest.raises(MisConfigurationError):
```
Comment (Borda) [inline tests/unittests/classification/test_accuracy.py:76]:
```suggestion
        with pytest.raises(ValueError):
```

File: tests/unittests/classification/test_accuracy.py [hunk 9]
```diff
@@ -60,6 +66,53 @@ def _sklearn_accuracy_binary(preds, target, ignore_index, multidim_average):
...
+                },
+            )
+
+    def test_accuracy_functional(self):
+        preds, target = _input_binary
+        task = "NotValidTask"
+        ignore_index = None
+        multidim_average = "global"
+
+        with pytest.raises(MisConfigurationError):
```
Comment (Borda) [inline tests/unittests/classification/test_accuracy.py:99]:
```suggestion
        with pytest.raises(ValueError):
```

File: tests/unittests/classification/test_accuracy.py [hunk 10]
```diff
@@ -60,6 +66,53 @@ def _sklearn_accuracy_binary(preds, target, ignore_index, multidim_average):
         return np.stack(res)
 
 
+class TestAccuracy(MetricTester):
```
Comment (Borda) [inline tests/unittests/classification/test_accuracy.py:69]: let's not instantiate full `MetricTester` class just for testing two specific cases

**Review by Borda — CHANGES_REQUESTED** [review]

File: tests/unittests/classification/test_accuracy.py (hunk 10 above)
Comment (JustinGoheen) [reply tests/unittests/classification/test_accuracy.py]: implemented in https://github.com/Lightning-AI/metrics/pull/1448/commits/199a71afb30aceee4e82c43dd0b6a7a7f9486cec

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
just the typing :)

**Comment by JustinGoheen** [comment]
@Borda I plan to work on the typing next week.

**Review by justusschock — APPROVED** [review]

**Comment by JustinGoheen** [comment]
@Borda @justusschock I decided to update the return type to `Optional[Tensor]` because the final case returns None. The pre-commit-ci bot added the return value 4 weeks ago, and that caused mypy to fail for the return value.

Is it better to have updated the expected return type, or to update the CI bot to not automate this type of update?

Screenshot

File: src/torchmetrics/functional/classification/accuracy.py [hunk 11]
```diff
@@ -378,12 +378,12 @@ def accuracy(
...
     num_labels: Optional[int] = None,
-    average: Optional[Literal["micro", "macro", "weighted", "none"]] = "micro",
-    multidim_average: Optional[Literal["global", "samplewise"]] = "global",
+    average: Literal["micro", "macro", "weighted", "none"] = "micro",
+    multidim_average: Literal["global", "samplewise"] = "global",
     top_k: Optional[int] = 1,
     ignore_index: Optional[int] = None,
     validate_args: bool = True,
-) -> Tensor:
+) -> Optional[Tensor]:
```
Comment (Borda) [inline src/torchmetrics/functional/classification/accuracy.py:386]: I think it shall always return only tensor r raise exception...
the same you may say about class with return None, this maybe NOne is confusing for user, and also any typing

**Review by Borda — CHANGES_REQUESTED** [review]

File: src/torchmetrics/functional/classification/accuracy.py (hunk 11 above)
Comment (JustinGoheen) [reply src/torchmetrics/functional/classification/accuracy.py:386]: are you okay with me adding `type: ignore[return-value]` to ignore the error at that line?

Comment (Borda) [reply src/torchmetrics/functional/classification/accuracy.py:386]: I think it is better option from the two

Comment (JustinGoheen) [reply src/torchmetrics/functional/classification/accuracy.py]: implemented in https://github.com/Lightning-AI/torchmetrics/pull/1448/commits/82800cdebc71d5c976f8652ce7529657e5149910

File: src/torchmetrics/functional/classification/accuracy.py [hunk 12]
```diff
@@ -409,18 +409,25 @@ def accuracy(
...
-        assert isinstance(num_labels, int)
+        if not isinstance(num_labels, int):
+            raise ValueError(
+                f"Optional arg `num_labels` must be type `int` when task is {task}. Got {type(num_labels)}"
+            )
         return multilabel_accuracy(
             preds, target, num_labels, threshold, average, multidim_average, ignore_index, validate_args
         )
-    raise ValueError(f"Not handled value: {task}")  # this is for compliant of mypy
+    return None  # type: ignore[return-value]
```
Comment (Borda) [inline src/torchmetrics/functional/classification/accuracy.py:433]:
```suggestion
    raise ValueError(f"Not handled value: {task}") # this is for compliant of mypy
```

File: src/torchmetrics/classification/accuracy.py [hunk 13]
```diff
@@ -480,16 +480,25 @@ def __new__(
...
+                raise ValueError(f"Optional arg `top_k` must be type `int` when task is {task}. Got {type(top_k)}")
             return MulticlassAccuracy(num_classes, top_k, average, **kwargs)
         if task == ClassificationTask.MULTILABEL:
-            assert isinstance(num_labels, int)
+            if not isinstance(num_labels, int):
+                raise ValueError(
+                    f"Optional arg `num_labels` must be type `int` when task is {task}. Got {type(num_labels)}"
+                )
             return MultilabelAccuracy(num_labels, threshold, average, **kwargs)
         return None
```
Comment (Borda) [inline src/torchmetrics/classification/accuracy.py:504]:
```suggestion
        raise ValueError(f"Not handled value: {task}") # this is for compliant of mypy
```

**Review by Borda — APPROVED** [review]

**Pull request merged**